src/psyclone/psyir/transformations/inline_trans.py

```diff
@@ -0,0 +1,478 @@
+        visitor = FortranWriter()
+        for dummy_arg, actual_arg in zip(routine_table.argument_list,
+                                         node.children):
+            # TODO #1799 this really needs the `datatype` method that is
+            # planned for all reference nodes.
+            if isinstance(actual_arg, ArrayReference):
+                rank = 0
+                for idx, expr in enumerate(actual_arg.indices):
+                    if isinstance(expr, Range):
+                        rank += 1
+                        # TODO #924 add a method to Range that returns the
+                        # PSyIR expression for the number of elements it
+                        # contains. This can then be compared with the shape of
+                        # the dummy argument.
+                        if not actual_arg.is_full_range(idx):
+                            raise TransformationError(LazyString(
+                                lambda: f"Cannot inline routine "
+                                f"'{routine.name}' because argument "
+                                f"'{visitor(actual_arg)}' is "
+                                f"an array subsection (TODO #924)."))
+                if isinstance(dummy_arg.datatype, ScalarType):
+                    dummy_rank = 0
+                else:
+                    dummy_rank = len(dummy_arg.datatype.shape)
+                if rank != dummy_rank:
+                    raise TransformationError(LazyString(
+                        lambda: f"Cannot inline routine '{routine.name}' "
+                        f"because it reshapes an argument: actual argument "
+                        f"'{visitor(actual_arg)}' has rank {rank} but the "
+                        f"corresponding dummy argument, '{dummy_arg.name}', "
+                        f"has rank {len(dummy_arg.datatype.shape)}"))
+
+    @staticmethod
+    def _find_routine(call_node):
+        '''
+        Searches for the definition of the routine that is being called by
+        the supplied Call.
+
+        Currently only supports routines that are present in the
+        same source file - TODO #924.
+
+        :returns: the PSyIR for the target routine.
+        :rtype: :py:class:`psyclone.psyir.nodes.Routine`
+
+        :raises TransformationError: if the RoutineSymbol is not local.
+        :raises InternalError: if the routine symbol is local but the \
+            definition cannot be found.
+        '''
+        name = call_node.routine.name
+        routine_sym = call_node.scope.symbol_table.lookup(name)
+        if not routine_sym.is_local:
+            raise TransformationError(
+                f"Routine '{name}' is imported and therefore cannot currently "
+                f"be inlined - TODO #924.")
+        table = routine_sym.find_symbol_table(call_node)
+        for routine in table.node.walk(Routine):
+            if routine.name == name:
+                return routine
+
+        raise InternalError(
```

Comment:
I wonder if this should be a TransformationError instead, things like recursive, pure or elemental keywords will put routines in CodeBlocks and not be found here, but it is not necessarily invalid PSyIR.

Reply:
Oh yes, I hadn't thought of that.